### ISSUE TITLE
Rework error classification logic in xrefs.go [phab:D2263]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2727]
_Created by **danielmoy-google** at 2018-04-24 00:18:52_: 

Because the upstream status returned doesn't always switch well, instead resort to really lame string comparison logic.

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-04-24 00:21:44_:

Hmm I don't quite like this either.

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-04-24 00:23:50_:

  - Check proper codes as well as string matching in xrefs error handling.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-24 02:33:51_:

At this point you already know that `code == codes.Unknown` by construction.

But if the gRPC status package doesn't know how to recognize context errors, I think we should at least file that as a bug upstream before we try to work around it by string matching. That's quite brittle, even for these fairly-specific error types.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-24 02:33:51_:

This might read a little nicer as a switch:

```
switch code := status.Code(err); code {
case codes.Canceled:
   …
case codes.DeadlineExceeded:
   …
case codes.Unknown:  // or maybe just default: ?
   …
}

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-24 03:40:51_:

Ah yea this is a bad commit, I'm halfway between a refactor here ((earlier comment)?id=5883 has the proper logic).

For the upstream request, will sync in person tomorrow since there's probably some internal bits that might make it a bit confusing who to chase for this.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-24 03:40:52_:

Yea I wasn't sure whether to repeat the return statements and separate the logic, or to clutter the logic and have single return statements.

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-04-24 03:40:57_:

  - prefer switch

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-04-25 18:07:21_:

PTAL

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-25 18:56:15_:

Is this always a ticket error? At this point we know the code is Unknown, so we could just propagate the original error I think?

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-25 19:25:23_:

If we propagate the original error, we're back to the original behavior, where we just return 13/internal error, which imo is worse.  I'd rather return something that explicitly says "we looked at this and didn't know" and attach some info, rather than just return 13/internal error.

The interface is always called with a ticket, at least currently.  If we ever don't, I think it'd probably be fine to drop the ticket, it's not *hugely* useful.  For now, kinda nice to have it in there imo.


That was my reasoning at least, but I have no super strong opinions either way on this.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2018-04-25 19:38:54_:

Sorry, my comment was too vague. I didn't mean we shouldn't return a status—only that we should not additionally decorate the error string we got. Hence: `return status.Error(code, err.Error())` or similar.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-04-25 20:20:56_:

ah gotcha, sure I guess that's reasonable.  We didn't get a ton of benefit from ticket, so w/e.  (Also in theory ticket is logged upstream in dremel, though sometimes that falls apart)

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-04-25 20:24:32_:

  - Drop the ticket.

**Comment by kythe** [comment]
_**fromberger** accepted this revision at 2018-04-25 21:55:57_

**Comment by kythe** [comment]
_**danielmoy-google** closed via commit 1d65895586b1475b2f3104778c74545e3bad4e15 at 2018-04-25 21:56:51_
